### PR TITLE
Better manage shape_crs and raster_crs in subset_shape

### PR DESCRIPTION
**Pull request by RondeauG**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
  - This PR fixes issue #xyz
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [x] I have added my relevant user information to `AUTHORS.md`

### What kind of change does this PR introduce?: 

* Fixes bugs in `subset_shape`:
  * Changes the order of the checks in `subset_shape` to check that the CRS are compatible at the start of the function, instead of after `subset_bbox`.
  * `poly` is reprojected to WGS84. This ensures that both `lon/lat_bnds` (for `subset_bbox`) and `mask_2d` are computed with compatible units and projections.

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@RondeauG Can you add a few dependency pins? Specifically:
- `numpy <2.1`
- `cartopy >=0.23`

I think those will solve the build failures.

File: environment.yml [hunk 1]
```diff
@@ -5,14 +5,15 @@ dependencies:
  - python >=3.8,<3.13
  - flit
  - bottleneck >=1.3.1
+ - cartopy >=0.23
```
Comment (RondeauG) [inline environment.yml:8]:
```suggestion
 - cartopy >=0.23.0
```

File: pyproject.toml [hunk 2]
```diff
@@ -48,14 +48,15 @@ dynamic = ["description", "version"]
 dependencies = [
   "bottleneck >=1.3.1",
   # cf-xarray is differently named on conda-forge
+  "cartopy >=0.23",
```
Comment (RondeauG) [inline pyproject.toml:51]:
```suggestion
  "cartopy >=0.23.0",
```

**Review by Zeitsperre — APPROVED** [review]
Maybe just update the changelog to mention the dependency adjustments and we're good to go!

**Pull request merged**